### ISSUE TITLE
Formatting `@doc` macro with trailing newline errors

### ISSUE THREAD

**Issue by devmotion** [domluna/JuliaFormatter.jl#555]
It seems there's a problem with `@doc` + trailing newlines:
```julia
julia> using JuliaFormatter

julia> format_text("@doc(filter)", BlueStyle())
"@doc(filter)"

julia> format_text("@doc filter", BlueStyle())
"@doc filter"

julia> format_text("@doc(Base.filter)", BlueStyle())
"@doc(Base.filter)"

julia> format_text("@doc Base.filter", BlueStyle())
"@doc Base.filter"

julia> format_text(
           raw"""
           @doc(filter)
           """,
           BlueStyle(),
       )
ERROR: Parsing error for input occurred on line 1, offset: 11
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] format_text(text::String, style::BlueStyle, opts::JuliaFormatter.Options)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:536
 [3] #format_text#206
   @ ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:530 [inlined]
 [4] format_text(text::String, style::BlueStyle)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:528
 [5] top-level scope
   @ REPL[11]:1

julia> format_text(
           raw"""
           @doc filter
           """,
           BlueStyle(),
       )
ERROR: Parsing error for input occurred on line 1, offset: 4
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] format_text(text::String, style::BlueStyle, opts::JuliaFormatter.Options)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:536
 [3] #format_text#206
   @ ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:530 [inlined]
 [4] format_text(text::String, style::BlueStyle)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:528
 [5] top-level scope
   @ REPL[12]:1

julia> format_text(
           raw"""
           @doc(Base.filter)
           """,
           BlueStyle(),
       )
ERROR: Parsing error for input occurred on line 1, offset: 16
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] format_text(text::String, style::BlueStyle, opts::JuliaFormatter.Options)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:536
 [3] #format_text#206
   @ ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:530 [inlined]
 [4] format_text(text::String, style::BlueStyle)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:528
 [5] top-level scope
   @ REPL[13]:1

julia> format_text(
           raw"""
           @doc Base.filter
           """,
           BlueStyle(),
       )
ERROR: Parsing error for input occurred on line 1, offset: 10
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] format_text(text::String, style::BlueStyle, opts::JuliaFormatter.Options)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:536
 [3] #format_text#206
   @ ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:530 [inlined]
 [4] format_text(text::String, style::BlueStyle)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/z3N5I/src/JuliaFormatter.jl:528
 [5] top-level scope
   @ REPL[14]:1
```

I suspect that possibly this is an upstream bug in CSTParser.

**Comment by domluna** [comment]
yeah if for parsing error happens for the `input` then it's CSTParser related

**Comment by devmotion** [comment]
I opened an issue in the CSTParser repo: https://github.com/julia-vscode/CSTParser.jl/issues/328

**Comment by devmotion** [comment]
With CSTParser 3.3.2, `@doc` with trailing newline errors only if there are no parentheses:
```julia
julia> using JuliaFormatter

julia> format_text(
           raw"""
           @doc(filter)
           """,
           BlueStyle(),
       )
"@doc(filter)\n"

julia> format_text(
           raw"""
           @doc filter
           """,
           BlueStyle(),
       )
ERROR: Parsing error for input occurred on line 1, offset: 4
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] format_text(text::String, style::BlueStyle, opts::JuliaFormatter.Options)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/1mHrw/src/JuliaFormatter.jl:536
 [3] #format_text#206
   @ ~/.julia/packages/JuliaFormatter/1mHrw/src/JuliaFormatter.jl:530 [inlined]
 [4] format_text(text::String, style::BlueStyle)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/1mHrw/src/JuliaFormatter.jl:528
 [5] top-level scope
   @ REPL[9]:1

julia> format_text(
           raw"""
           @doc(Base.filter)
           """,
           BlueStyle(),
       )
"@doc(Base.filter)\n"

julia> format_text(
           raw"""
           @doc Base.filter
           """,
           BlueStyle(),
       )
ERROR: Parsing error for input occurred on line 1, offset: 10
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] format_text(text::String, style::BlueStyle, opts::JuliaFormatter.Options)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/1mHrw/src/JuliaFormatter.jl:536
 [3] #format_text#206
   @ ~/.julia/packages/JuliaFormatter/1mHrw/src/JuliaFormatter.jl:530 [inlined]
 [4] format_text(text::String, style::BlueStyle)
   @ JuliaFormatter ~/.julia/packages/JuliaFormatter/1mHrw/src/JuliaFormatter.jl:528
 [5] top-level scope
   @ REPL[11]:1
```

**Comment by devmotion** [comment]
The remaining errors are fixed in CSTParser 3.3.3.
